### PR TITLE
Add pImpl to `libdnf5::base::transaction_*` classes

### PR DESCRIPTION
**Pull request by kontura**
This includes package, group, environment and module.

This PR is into a new dnf5-5.2.0.0 branch to keep the main branch API compatible.

Another part for: https://github.com/rpm-software-management/dnf5/issues/1025

### CODE REVIEW

File: include/libdnf5/base/transaction_package.hpp [hunk 1]
```diff
@@ -43,70 +42,60 @@ class TransactionPackage {
...
-    State state{State::STARTED};
-    std::optional<std::string> reason_change_group_id;
-
-    std::vector<rpm::Package> replaces;
-    std::vector<rpm::Package> replaced_by;
+        const libdnf5::rpm::Package & pkg, Action action, Reason reason, const std::optional<std::string> & group_id);
+
+    void set_reason(Reason value) noexcept;
+    void set_state(State value) noexcept;
+    void replaced_by_append(rpm::Package && pkg);
```
Comment (jan-kolarik) [inline include/libdnf5/base/transaction_package.hpp:94]: This (and the one below) is also used only in impl right? Probably not needed in the public API.

File: libdnf5/base/transaction_environment.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            p_impl = std::make_unique<Impl>(*mpkg.p_impl);
+        }
+    }
+
+    return *this;
+}
+TransactionEnvironment & TransactionEnvironment::operator=(TransactionEnvironment && mpkg) noexcept = default;
+
+TransactionEnvironment::Impl::Impl(
+    libdnf5::comps::Environment grp, Action action, Reason reason, const bool with_optional)
```
Comment (jan-kolarik) [inline libdnf5/base/transaction_environment.cpp:59]: I know the `const bool` was present in the original code, but the `const` doesn't have any meaning there, or does it? Also on L66.

File: include/libdnf5/base/transaction_environment.hpp [hunk 3]
```diff
@@ -36,46 +35,44 @@ class TransactionEnvironment {
...
-    TransactionEnvironment(
-        const libdnf5::comps::Environment & grp, Action action, Reason reason, const bool with_optional)
-        : environment(grp),
-          action(action),
-          reason(reason),
-          with_optional(with_optional) {}
+    TransactionEnvironment(const libdnf5::comps::Environment & grp, Action action, Reason reason, bool with_optional);
 
-    void set_state(State value) noexcept { state = value; }
+    void set_state(State value) noexcept;
```
[the remark this answers is not in the document]
Comment (kontura) [reply include/libdnf5/base/transaction_environment.hpp:72]: It is actually not even used now, I will remove it.

File: include/libdnf5/base/transaction_package.hpp (hunk 1 above)
Comment (kontura) [reply include/libdnf5/base/transaction_package.hpp:94]: It will require a new private header but I can do that.

Comment (kontura) [reply include/libdnf5/base/transaction_package.hpp]: I have moved also the `set_reason(...)` to the Impl.

**Review by jan-kolarik — APPROVED** [review]
LGTM, thanks!

**Pull request merged**